app/src/main/java/com/google/android/samples/socialite/ui/player/preloadmanager/PreloadManagerWrapper.kt

```diff
@@ -0,0 +1,166 @@
+
+    private var currentPlayingIndex = C.INDEX_UNSET
+
+    // List of all items in our current list of media items to be rendered on the UI
+    private var mediaItemsList = listOf<TimelineMediaItem>()
+
+    // Defines when to start preloading next items w.r.t current playing item
+    private val itemsRemainingToStartNextPreloading = 2
+
+    /** Builds a preload manager instance with default parameters. Preload manager should use the same looper and load control as the player */
+    companion object {
+        fun build(
+            playbackLooper: Looper,
+            loadControl: DefaultLoadControl,
+            context: Context,
+        ): PreloadManagerWrapper {
+            val trackSelector = DefaultTrackSelector(context)
+            trackSelector.init({}, DefaultBandwidthMeter.getSingletonInstance(context))
+            val renderersFactory = DefaultRenderersFactory(context)
+            val preloadManager = DefaultPreloadManager(
+                PreloadStatusControl(),
+                DefaultMediaSourceFactory(context),
+                trackSelector,
+                DefaultBandwidthMeter.getSingletonInstance(context),
+                DefaultRendererCapabilitiesList.Factory(renderersFactory),
+                loadControl.allocator,
+                playbackLooper,
+            )
+            return PreloadManagerWrapper(preloadManager)
+        }
+    }
+
+    /** Add initial list of videos to the preload manager. */
+    fun init(mediaList: List<TimelineMediaItem>) {
+        if (mediaList.isEmpty()) {
+            return
+        }
+        setCurrentPlayingIndex(0)
+        setMediaList(mediaList)
+        preloadNextItems()
+    }
+
+    /** Sets the index of the current playing media. */
+    fun setCurrentPlayingIndex(currentPlayingItemIndex: Int) {
+        currentPlayingIndex = currentPlayingItemIndex
+        defaultPreloadManager.setCurrentPlayingIndex(currentPlayingIndex)
+        preloadNextItems()
+    }
+
+    /** Sets the list of media items to be played. Can be set as and when new data is loaded. */
+    private fun setMediaList(mediaList: List<TimelineMediaItem>) {
+        mediaItemsList = mediaList
+    }
+
+    /** Ensure that current playing item is in the middle of the preload Window . */
+    private fun preloadNextItems() {
+        var lastPreloadedIndex = 0
+        if (!preloadWindow.isEmpty()) {
+            lastPreloadedIndex = preloadWindow.last().second
+        }
+
+        if (lastPreloadedIndex - currentPlayingIndex <= itemsRemainingToStartNextPreloading) {
+            for (i in 1 until (preloadWindowMaxSize - itemsRemainingToStartNextPreloading)) {
+                addMediaItem(index = lastPreloadedIndex + i)
+                removeMediaItem()
+            }
+            // With invalidate, preload manager will internally sort the priorities of all the media items added to it, and trigger the preload from the most important one
+            defaultPreloadManager.invalidate()
```

Comment:
We should call `invalidate()` outside of the `if` condition, because anyway we need `invalidate()` as the current playing index has changed, or when the initial list of media is added we have to trigger the preload.

Reply:
Not sure if this comment is actually re-opened. I added a new one for visibility instead.